Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/CommonComponents/AdditionalFieldsWindow.tsx

```diff
@@ -55,12 +86,41 @@ function AdditionalFieldsWindow(props: IProps): JSX.Element {
                 <AdditionalFieldsTable
                     ID={props.ID}
                     Type={props.Type}
-                    SaveFieldsCallback={saveCallback}
-                    ResetFieldsCallback={resetCallback}
-                    SetChangedMessageList={setChanges}
-                    SetInvalidMessageList={setInvalidChanges}
-                    SetHasEditPermissions={setHasPermissions}
-                    HideExternal={false} />
+                    FieldValues={allValues}
+                    HideExternal={false}
+                    LoadState={state}
+                    SetValues={(value, field) => {
+                        const i = changes.findIndex(f => f.AdditionalFieldID === field.ID);
+                        const iCurrent = values.findIndex(f => f.AdditionalFieldID === field.ID);
+
+                        if (i >= 0 && iCurrent < 0)
+                            setChanges((c) => { const u = _.cloneDeep(c); u[i] = value; return u })
+                        else if (i >= 0 && values[iCurrent].Value != value.Value)
+                            setChanges((c) => { const u = _.cloneDeep(c); u[i] = value; return u })
+                        else if (i >= 0)
+                            setChanges((c) => { const u = _.cloneDeep(c); u.splice(i); return u })
+                        else if (iCurrent >= 0 && values[iCurrent].Value != value.Value)
+                            setChanges((c) => { const u = _.cloneDeep(c); u.push(value); return u })
+                        else if (iCurrent < 0)
+                            setChanges((c) => { const u = _.cloneDeep(c); u.push(value); return u })
+
+                        const iInvalid = invalidChanges.findIndex(item => item.FieldID == field.ID);
+                        const iChange = validChanges.findIndex(item => item.FieldID == field.ID);
+                        const valid = field.Type != 'integer' || (value.Value != null && IsInteger(value.Value));
+        
+                        if (valid && iInvalid >= 0)
+                            setInvalidChanges((c) => { const u = _.cloneDeep(c); u.splice(iInvalid); return u });
+                        if (!valid && iInvalid < 0) 
+                            setInvalidChanges((c) => { const u = _.cloneDeep(c); u.push({ FieldID: field.ID, Message: `Value for \'${field.FieldName}\' must be an integer.` }); return u });
+
+                        if (valid && iCurrent >= 0 && iChange < 0) 
+                            setValidChanges((c) => { const u = _.cloneDeep(c); u.push({ FieldID: field.ID, Message: `Changes to \'${field.FieldName}\' will be lost.`}); return u });
+                        if (!valid && iCurrent >= 0 && iChange >= 0) 
+                            setValidChanges((c) => { const u = _.cloneDeep(c); u.splice(iChange); return u });
+
```

Comment:
I think this is going to lead to potential issues down the line with the states of invalid changes, valid changes, and changes mismatching on what they say. Regenerating invalid/valid changes here should not be a big performance loss since we have to iterate the whole thing to clone it anyway. We could also leave that regeneration to a use effect so that it catches any changes in case this expands in the future.

Reply:
Except we can't regenerate `validChanges` or `invalidChanges` because they contain the `field.FieldName` and `changes` does not include the `FieldName` so we can only generate an `validChange` or `invaliudChange` when we have the `field` which is only the case for the `change` this is currently processing.